server/bin/pbench-server-prep-shim-002.py

```diff
@@ -151,6 +153,27 @@ def process_tb(config, logger, receive_dir, qdir_md5, duplicates, errors):
             nquarantined += 1
             continue
 
+        # Create a new dataset tracker in UPLOADING state, and add it to the
+        # database.
+        #
+        # NOTE: Technically, this particular workflow has no "UPLOADING" as
+        # the `pbench-server-prep-shim-002` command isn't invoked until the
+        # tarball and MD5 has been entirely uploaded by the agent via `ssh`;
+        # this method however can't be supported once we have authorized user
+        # ownership, and the model fits the server `PUT` method where an
+        # unexpected termination could leave a tarball in "Uploading" state.
+        #
+        # TODO: For now, if we're unable to create it for some reason,
+        # log it and move on. Need to decide how to handle this for real.
+        #
+        # TODO: We create_or_attach because we might hit backup before the
```

Comment:
Not sure how this can happen.  The backup task should only run from dispatch, and the prep-shim feeds dispatch.

Reply:
Happens all the time in the gold unittests, though I wrote the comment a bit more generically because at the time I hadn't considered the fact that prep_shim would have to finish to make the tarball visible to backup.

Reply:
That sounds like a problem with the way the unit tests are constructed.

Reply:
Well, in a way, yeah, you're right. I suppose this is my mistake because I wanted to minimize the depth of immersion in the gold unittest infrastructure. 

E.g., test-6.x runs pbench-backup-tarballs all by itself. That means there's nothing to set up the datasets in advance.

However, just as we package tarballs for the expected intake/incoming directory trees to prepare for the tests, to be fully clean I would have created pickled datasets in an Sqlite DB file that could be laid down in the right place before running pbench-backup-tarballs with all the expected datasets in the right state.

This obviously would be an enormous amount of setup overhead. I chose to compromise the integrity of the model by allowing these `attach_or_create` touchpoints that really shouldn't exist rather than going to all that work for the gold unittest environment we know we want to get rid of anyway.

So I guess the real issue is whether you want to challenge that lazy tradeoff and push me into doing it right? ;-)

Reply:
That is a good question.  Not sure what the right call is, but the more I consider this work, the more foundational it becomes and the more I'd like to consider making the `-server-prep-shim-` (and for that matter, the "satellite" integration) work using the `PUT` API.

Reply:
> That is a good question. Not sure what the right call is, but the more I consider this work, the more foundational it becomes and the more I'd like to consider making the `-server-prep-shim-` (and for that matter, the "satellite" integration) work using the `PUT` API.

I really don't see how that helps us much; ultimately we need everything to have a real username, and "shimming the shim" won't do that. Everything coming from a satellite or stuffed by ssh would be "public". And maybe that's OK for short-term 0.69 agent compatibility, but it's not a long term strategy.